packages/mobx/src/core/globalstate.ts

```diff
@@ -90,9 +90,9 @@ export class MobXGlobals {
 
     /**
      * Is it allowed to read observables at this point?
-     * Used to hold the state needed for `observableRequiresReaction`
+     * Used to hold the state needed for `observableRequiresReaction`/`computedRequiresReaction`
      */
-    allowStateReads = true
+    allowStateReads = false
```

Comment:
This might be a breaking change? By default we allow state reads outside reactive contexts?

Reply:
No. These flags, `allowStateReads` and `allowStateChanges`, says whether you're in a context, where reads/changes are allowed - or perhaps better where they are supposed to take place. They're basically `inDerivation`/`inAction` flags - they're switched on/off when entering/leaving these, regardless of whether it should warn or not. We use these flags instead of directly checking for `action`/`derivation`, because the mapping isn't always exactly 1:1. But really the point is to demark these places, not to directly control the warning - the warning is always only raised in combination with `requiresReaction`/`enforceAction`. They should always start with `false`, because there is no running action/derivation by default.

Reply:
Check, that adds up. I haven't been working on this for too long 😅.

Reply:
Turns out it's more complicated. I think it was originally designed as described, but it was maybe misunderstood and changed with the introduction of `autoAction`.
The thing is that `autoAction` is supposed to warn when you modify state inside derivation, regardless of `enforceAction` configuration. So `autoAction` assumes that if `allowStateChanges` is false, there will be a warning when setting state, no other conditions needed.
In order for this to work the `configure/resetGlobalState` was updated to synchronize `allowStateChanges` with `enforeAction` configuration.
But the actual check for the warning doesn't assume this behavior:
https://github.com/mobxjs/mobx/blob/13a222ea21bd55b7158d48a998846da608080ded/packages/mobx/src/core/derivation.ts#L140-L143
Notice there is a bug - the check doesn't respect `enforceActions: "never"` (enforceActions === false).
This bug is actually a reason why [the test for `autoAction` warning](https://github.com/mobxjs/mobx/blob/13a222ea21bd55b7158d48a998846da608080ded/packages/mobx/__tests__/v5/base/action.js#L603-L623) is passing - there is [a test above](https://github.com/mobxjs/mobx/blob/13a222ea21bd55b7158d48a998846da608080ded/packages/mobx/__tests__/v5/base/action.js#L308) that calls `mobx.configure({ enforceActions: "never" })`.
If you fix this bug, the `autoAction` warning (*Side effects like changing state are not allowed*) can never occur.